src/System.Security.Cryptography.Pkcs/src/Internal/Cryptography/Helpers.cs

```diff
@@ -89,34 +89,82 @@ public static X509Certificate2 TryFindMatchingCertificate(this X509Certificate2C
                 case SubjectIdentifierType.IssuerAndSerialNumber:
                     {
                         X509IssuerSerial issuerSerial = (X509IssuerSerial)(recipientIdentifier.Value);
-
-                        using (ClonedCertificates matches1 = certs.FindCertificates(X509FindType.FindBySerialNumber, issuerSerial.SerialNumber))
+                        string serialNumber = issuerSerial.SerialNumber;
+                        string issuer = issuerSerial.IssuerName;
+                        foreach (X509Certificate2 candidate in certs)
                         {
-                            using (ClonedCertificates matches2 = matches1.FindCertificates(X509FindType.FindByIssuerDistinguishedName, issuerSerial.IssuerName))
-                            {
-                                // Desktop compat: We do not complain about multiple matches. Just take the first one and ignore the rest.
-                                return matches2.Any ? matches2.ClaimResult() : null;
-                            }
+                            if (candidate.SerialNumber == serialNumber && candidate.Issuer == issuer)
```

Comment:
X509Certificate has a GetSerialNumber() method that returns a serial as a byte[] array. Probably better to compare that (that way, the normalization agreement is all in the PKCS package since it’s the one that creates the stringized version in X509IssuerSerial to begin with.)


From: Jeremy Barton [mailto:notifications@github.com]
Sent: Thursday, May 5, 2016 2:21 PM
To: dotnet/corefx <corefx@noreply.github.com>
Cc: Atsushi Kanamori <atsushik@microsoft.com>; Author <author@noreply.github.com>
Subject: Re: [dotnet/corefx] Fix EnvelopeCms Decrypt() failure on collection-imported PFXs. (#8294)


In src/System.Security.Cryptography.Pkcs/src/Internal/Cryptography/Helpers.cs<https://github.com/dotnet/corefx/pull/8294#discussion_r62257220>:

>                          {

> -                            using (ClonedCertificates matches2 = matches1.FindCertificates(X509FindType.FindByIssuerDistinguishedName, issuerSerial.IssuerName))

> -                            {

> -                                // Desktop compat: We do not complain about multiple matches. Just take the first one and ignore the rest.

> -                                return matches2.Any ? matches2.ClaimResult() : null;

> -                            }

> +                            if (candidate.SerialNumber == serialNumber && candidate.Issuer == issuer)

SerialNumber has a lot of fuzzy logic here.

Potentially important:

  *   The hex string is matched case insensitively.
     *   X509Certificate2.SerialNumber is all uppercased
  *   Any spaces, colons, or hyphens are ignored
  *   The number is also interpreted as a decimal value. ("1024" matches either { 0x04, 0x00 } or {0x10, 0x24})

Less important, but still compat-breaking:

  *   Any non-hex character is ignored. "q1" matches { 0x01 }.
  *   Any non-decimal character is ignored for the decimal parsing ("13f" will match { 0x01, 0x3f } and { 0x0d } (13)

Since we're responsible for decoding the serial number, and the type isn't externally constructable, this is probably fine, though. (But maybe it's worth a comment that since we build both strings from byte[] we know the values are normalized)

—
You are receiving this because you authored the thread.
Reply to this email directly or view it on GitHub<https://github.com/dotnet/corefx/pull/8294/files/2772d787f16f24ef6b1f1c5de097aae9a565712f#r62257220>